core/src/main/java/com/graphhopper/routing/util/FerrySpeedCalculator.java

```diff
@@ -8,65 +8,58 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(FerrySpeedCalculator.class);
     private final double speedFactor;
-    private final double unknownSpeed, longSpeed, shortSpeed, maxSpeed;
+    private final double unknownSpeed, maxSpeed;
 
-    public FerrySpeedCalculator(double speedFactor, double maxSpeed, double longSpeed, double shortSpeed, double unknownSpeed) {
+    public FerrySpeedCalculator(double speedFactor, double maxSpeed, double unknownSpeed) {
         this.speedFactor = speedFactor;
         this.unknownSpeed = unknownSpeed;
-        this.longSpeed = longSpeed;
-        this.shortSpeed = shortSpeed;
         this.maxSpeed = maxSpeed;
     }
 
     /**
      * Special handling for ferry ways.
      */
     public double getSpeed(ReaderWay way) {
-        // During the reader process we have converted the duration value into an artificial tag called "duration:seconds".
+        // todo: we should re-consider whether we should deal with ferries by determining (only) a speed at all.
+        //       for example this way we cannot account for waiting times for short ferries (speed is too low) or
+        //       'fast' ferries that exceed the max speed for e.g. the foot encoder. Maybe we could add an additional
+        //       encoded value that stores the ferry waiting time (maybe just short/medium/long or so) and add a few
+        //       more possible speed values to account for high ferry speeds for the slow vehicles like foot.
+
+        // During the reader process we have added the artificial "road_distance" tag and converted the duration tag to
+        // an artificial tag called "duration:seconds"
+        double distanceInKm = way.getTag("road_distance", -1.0) / 1000;
+        if (distanceInKm < 0)
+            throw new IllegalStateException("The artificial 'road_distance' tag is missing for way: " + way.getId());
         Long duration = way.getTag("duration:seconds", 0L);
-        // seconds to hours
         double durationInHours = duration / 60d / 60d;
-        // Check if our graphhopper specific artificially created estimated_distance way tag is present
-        Number estimatedLength = way.getTag("estimated_distance", null);
         if (durationInHours > 0) {
-            if (estimatedLength != null) {
```

Comment:
`estimatedLength != null` can never happen.

Reply:
I only just understood this now: Before #2448 `estimatedLength == null` **could** happen when the first/last OSM node of a ferry way was missing:

https://github.com/graphhopper/graphhopper/blob/f95397b7dbcb186de7cafb735af8275abd6122d0/core/src/main/java/com/graphhopper/reader/osm/OSMReader.java#L206-L212

And the reason we have to consider the distance of the entire way is that 1) the duration tag refers to the duration of the entire way and 2) ferry ways sometimes intersect and are thus split into several edges for which we cannot directly use the duration tag.